[EPMEDU-968]: Open the feeding point details from the Favorites page

ArtemKleinschmidt opened the pull request:
Open the feeding point details from the Favorites page

Moved HomeBottomSheet to foundation
Moved FeedingPointSheetContent and dependencies to shared
Added feeding details dialog to the favourites screen without modifications
Added "go to map" link
Extracted alpha calculation logic to make it shared


[device-2022-11-10-144435.webm](https://user-images.githubusercontent.com/32295746/201096418-2d5fd65b-a8dc-4631-8d19-d411b653c3c7.webm)

ArtemKleinschmidt committed:
[EPMEDU-1068] Open the feeding point details from the Favorites page

Moved HomeBottomSheet to foundation
Moved FeedingPointSheetContent and dependencies to shared
Added feeding details dialog to the favorites screen without modifications

ArtemKleinschmidt committed:
Make feed button unavailable is an animal is fed
Added "go to map" link
Extracted alpha calculation logic to make it shared

ArtemKleinschmidt committed:
Code smells fix

carlossantiagomenendez commented on feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/FavouritesScreenUI.kt:
```diff
@@ -1,31 +1,72 @@
+
+@Composable
+private fun ScreenScaffold(
+    bottomSheetState: HomeBottomSheetState,
```
Hi! The bottom sheet looks sharp, the only issue I see is that is displaying some bottom padding making it different from mapBottomSheet so please remove the '.bottomBarPadding()'. And maybe put contentAlpha and buttonAlpha in order?
Nice job!

carlossantiagomenendez commented on shared/feature/feeding/build.gradle.kts:
```diff
@@ -0,0 +1,28 @@
+    implementation(projects.library.resources)
+    implementation(projects.library.geolocation)
+
+    implementation(libs.androidx.datastore)
```
I think there is no need for datastore, material.dialog.core or kapt plugin.
Please delete them.

ArtemKleinschmidt committed:
Code review comments fixes

Fileek commented on shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/presentation/ui/FeedingPointSheet.kt:
```diff
@@ -13,27 +13,29 @@ import androidx.compose.ui.draw.clip
+fun FeedingPointSheetContent(
+    feedingPoint: FeedingPointModel,
     contentAlpha: Float,
+    showOnMap: Boolean = false,
```
Can you rename it to isShowOnMapVisible?

Fileek commented on shared/feature/feeding/src/main/AndroidManifest.xml:
```diff
@@ -0,0 +1,3 @@
+<?xml version="1.0" encoding="utf-8"?>
+<manifest>
+</manifest>
```
You can collapse it to `<manifest/>`

Fileek reviewed (changes requested):
Please resolve conflicts with the develop branch

ArtemKleinschmidt committed:
Code review comments fixes

ArtemKleinschmidt committed:
Merge branch 'develop' into feature/EPMEDU-968_open_the_feeding_point_details_from_the_favorites_page

# Conflicts:
#	feature/home/build.gradle.kts
#	feature/home/src/main/java/com/epmedu/animeal/home/presentation/HomeScreenUI.kt
#	feature/home/src/main/java/com/epmedu/animeal/home/presentation/viewmodel/HomeState.kt
#	feature/home/src/main/java/com/epmedu/animeal/home/presentation/viewmodel/HomeViewModel.kt
#	library/resources/src/main/res/values/strings.xml
#	settings.gradle.kts

ArtemKleinschmidt committed:
Fixed errors after merge

ArtemKleinschmidt committed:
Added feed confirmation dialog to the favourites screen

ArtemKleinschmidt committed:
Code smell fix

Fileek reviewed (changes requested):
Please check the designs. There is no collapsed state and bottom sheet should cover whole screen

ArtemKleinschmidt committed:
Make feeding dialog fullscreen for the favorites screen

ArtemKleinschmidt commented:
> 

Hey @Fileek  I've upda
[device-2022-11-21-141030.webm](https://user-images.githubusercontent.com/32295746/203050761-6a27f66b-5bd3-4ecc-b24a-a7828660d5ad.webm)
ted the dialog to be fullscreen on the favorites page. Please check the code and the attached video

ArtemKleinschmidt committed:
Code smells fix

ArtemKleinschmidt committed:
minor UI update

Fileek commented:
> > 
> 
> Hey @Fileek I've updated the dialog to be fullscreen on the favorites page. Please check the code and the attached video [device-2022-11-21-195113.webm](https://user-images.githubusercontent.com/32295746/203125750-5744fea9-6004-400d-9c49-7d07aab4df6c.webm)

Good job. But can I ask you to hide the bottom bar when the bottom sheet animation starts and not when it ends?

ArtemKleinschmidt committed:
Hide bottom bar when animation starts

ArtemKleinschmidt commented:
> > > 
> > 
> > 
> > Hey @Fileek I've updated the dialog to be fullscreen on the favorites page. Please check the code and the attached video [device-2022-11-21-195113.webm](https://user-images.githubusercontent.com/32295746/203125750-5744fea9-6004-400d-9c49-7d07aab4df6c.webm)
> 
> Good job. But can I ask you to hide the bottom bar when the bottom sheet animation starts and not when it ends?

Updated the animation 
[device-2022-11-24-142557.webm](https://user-images.githubusercontent.com/32295746/203785363-b4f56777-6006-43fb-a98f-f94e1e65d526.webm)

Fileek commented on library/foundation/src/main/java/com/epmedu/animeal/foundation/dialog/bottomsheet/FeedingPointActionButton.kt:
```diff
@@ -0,0 +1,38 @@
+package com.epmedu.animeal.foundation.dialog.bottomsheet
```
It is a part of the feeding feature, it shouldn't be in the library. Please move it to `feeding.presentation.ui`

Fileek commented on library/foundation/src/main/java/com/epmedu/animeal/foundation/dialog/bottomsheet/HomeBottomSheet.kt:
```diff
@@ -212,4 +228,33 @@ private fun Modifier.bottomSheetSwipeable(
+
+@OptIn(ExperimentalMaterialApi::class)
+@Composable
+fun HomeBottomSheetState.contentAlphaButtonAlpha(): Pair<Float, Float> {
```
Could you replace `Home` word with `Animeal` or remove it in the composables which you've extracted?

Fileek commented on library/foundation/src/main/java/com/epmedu/animeal/foundation/dialog/bottomsheet/HomeBottomSheet.kt:
```diff
@@ -118,6 +119,7 @@ fun HomeBottomSheetLayout(
     sheetElevation: Dp = ModalBottomSheetDefaults.Elevation,
     sheetBackgroundColor: Color = MaterialTheme.colors.surface,
     sheetContentColor: Color = contentColorFor(sheetBackgroundColor),
+    isShownStateEnabled: Boolean = true,
```
Could you rename it to something like `isAlwaysExpanded` or something similar to make this boolean more clear?

Fileek commented on feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/FavouritesScreen.kt:
```diff
@@ -1,14 +1,42 @@
 package com.epmedu.animeal.favourites
```
Please move it to `presentation` package

Fileek commented on feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/FavouritesScreenEvent.kt:
```diff
@@ -2,5 +2,8 @@ package com.epmedu.animeal.favourites
 
 internal sealed class FavouritesScreenEvent(open val id: Int = -1) {
```
Please move it to `presentation` package

Fileek commented on feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/FavouritesScreenUI.kt:
```diff
@@ -1,50 +1,134 @@
 package com.epmedu.animeal.favourites
```
Same as above

Fileek commented on feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/FavouritesState.kt:
```diff
@@ -1,5 +1,9 @@
 package com.epmedu.animeal.favourites
```
Please move it to `presentation.viewmodel` package

Fileek commented on feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/FavouritesViewModel.kt:
```diff
@@ -13,7 +15,8 @@ import javax.inject.Inject
 internal class FavouritesViewModel @Inject constructor(
```
Please move it to `presentation.viewmodel` package

Fileek commented on shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/presentation/model/FeedStatus.kt:
```diff
@@ -0,0 +1,10 @@
+import com.epmedu.animeal.feeding.data.model.enum.AnimalState
+import com.epmedu.animeal.foundation.common.FeedStatus
+
+fun AnimalState.toFeedStatus(): FeedStatus =
```
Could you please move `FeedStatus` enum to here as well and `FeedStatus` composable to the ui package? Also remove one of the `FeedStatus` composables, one is in the library and second one is in `FeedingPointSheet`

Fileek commented on feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/FavouritesViewModel.kt:
```diff
@@ -25,11 +28,18 @@ internal class FavouritesViewModel @Inject constructor(
+                updateState { copy(showingWillFeedDialog = true) }
             }
         }
+        viewModelScope.launch { sendEvent(event) }
```
Screen events are intended to be sent by screen. You should handle them here and react correspondingly, but you can't just send them back. Create a new sealed class for events from viewmodel - `FavouritesEvent` or `FavouritesViewModelEvent` and put it into `presentation.viewmodel` package

Fileek commented on feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/FavouritesScreenEvent.kt:
```diff
@@ -2,5 +2,8 @@ package com.epmedu.animeal.favourites
 internal sealed class FavouritesScreenEvent(open val id: Int = -1) {
     data class FeedSpotSelected(override val id: Int) : FavouritesScreenEvent(id)
-    data class FeedSpotRemove(override val id: Int) : FavouritesScreenEvent(id)
+    data class FeedSpotChanged(override val id: Int, val isFavorite: Boolean) :
```
Why did you rename this event and add a boolean if you use it the same way? I guess we have to ask BAs if we can add feeding points back to favorites using bottom sheet

Fileek commented on feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/FavouritesScreenUI.kt:
```diff
@@ -102,23 +201,34 @@ private fun FavouritesScreenPreview() {
                 )
-            )
+            ),
+            rememberHomeBottomSheetState(HomeBottomSheetValue.Hidden)
```
Don't use remember in preview, it's static anyway

ArtemKleinschmidt replied:
@Fileek  State is an obligatory parameter, need to pass something here

Fileek replied:
Just create it as an instance of the class

Fileek commented on feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/FavouritesScreenUI.kt:
```diff
@@ -102,23 +201,34 @@ private fun FavouritesScreenPreview() {
                 emptyList()
-            )
+            ),
+            rememberHomeBottomSheetState(HomeBottomSheetValue.Hidden)
```
Same as above

Fileek commented on library/foundation/src/main/java/com/epmedu/animeal/foundation/dialog/bottomsheet/HomeBottomSheet.kt:
```diff
@@ -1,6 +1,7 @@
-package com.epmedu.animeal.home.presentation.ui
+package com.epmedu.animeal.foundation.dialog.bottomsheet
```
It is a part of the feeding feature, it shouldn't be in the library. Please move it to feeding.presentation.ui

ArtemKleinschmidt replied:
@Fileek  it is not a part of feeding feature. It is a custom bottom sheet dialog, that can be used with custom content anywhere in the app

Fileek commented on feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/FavouritesScreen.kt:
```diff
@@ -1,14 +1,42 @@
+        viewModel.events.collect {
+            when (it) {
+                is FavouritesScreenEvent.FeedSpotSelected -> {
+                    launch {
```
Why do you launch child coroutine here, but not below?

Fileek commented on shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/presentation/ui/FeedingPointSheet.kt:
```diff
@@ -13,27 +13,30 @@ import androidx.compose.ui.draw.clip
+fun FeedingPointSheetContent(
+    feedingPoint: FeedingPointModel,
     contentAlpha: Float,
+    expandToFullScreen: Boolean = false,
```
You should use modifier and pass it to the column instead of Spacer depended on this boolean

ArtemKleinschmidt committed:
Code review comments fixes

ArtemKleinschmidt committed:
Fixed package structure
Fixed events and state usage
Fixed like button behavior on the favourites screen

ArtemKleinschmidt committed:
code smells fix

ArtemKleinschmidt committed:
Merge develop

Fileek commented on shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/presentation/ui/FeedStatusItem.kt:
```diff
@@ -48,11 +44,18 @@ fun FeedStatus(
-    GREEN(R.string.feed_status_green, R.drawable.ic_face_happy, CustomColor.StatusGreen),
-    RED(R.string.feed_status_red, R.drawable.ic_face_upset, CustomColor.StatusRed);
+@Composable
+@Preview
```
Please use AnimealPreview

Fileek commented on shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/presentation/ui/FeedStatusItem.kt:
```diff
@@ -48,11 +44,18 @@ fun FeedStatus(
+
+@Composable
+@Preview
+fun FeedStatusItemRedPreview() {
```
It's better to put different previews of one small ui component in one preview using column and dividers

Fileek commented on shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/presentation/ui/FeedingPointSheet.kt:
```diff
@@ -18,38 +18,42 @@ import androidx.compose.ui.text.style.TextOverflow
                 end = 24.dp,
-            ),
+            )
+            .then(modifier),
```
You should use this modifier in the beginning as the receiver of the chain

ArtemKleinschmidt committed:
Code review fixes

ArtemKleinschmidt committed:
Merge remote-tracking branch 'origin/develop' into feature/EPMEDU-968_open_the_feeding_point_details_from_the_favorites_page

# Conflicts:
#	feature/home/src/main/java/com/epmedu/animeal/home/presentation/HomeScreenUI.kt
#	feature/home/src/main/java/com/epmedu/animeal/home/presentation/ui/MapboxMap.kt

ArtemKleinschmidt committed:
Code smell fix

Fileek commented on library/foundation/src/main/java/com/epmedu/animeal/foundation/dialog/bottomsheet/AnimealBottomSheet.kt:
```diff
@@ -1,6 +1,7 @@
-package com.epmedu.animeal.home.presentation.ui
+package com.epmedu.animeal.foundation.dialog.bottomsheet
```
Shouldn't be in `dialog`

Fileek commented on feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/presentation/viewmodel/FavouritesState.kt:
```diff
@@ -0,0 +1,10 @@
+import com.epmedu.animeal.feeding.domain.model.FeedingPoint
+
+data class FavouritesState(
+    val favouritesSnapshot: List<FeedingPoint> = emptyList(),
```
Looks like you don't need this list on UI. Move it to the viewmodel as a private variable instead

Fileek commented on feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/presentation/viewmodel/FavouritesViewModel.kt:
```diff
@@ -0,0 +1,77 @@
+
+    private fun updateSnapshot(event: FavouritesScreenEvent.FeedSpotChanged): MutableList<FeedingPoint> {
+        val snapshot = mutableListOf<FeedingPoint>()
+        state.favouritesSnapshot.forEach { feedingPoint ->
```
Can be simplified with map operator

Fileek commented on feature/tabsflow/favourites/src/main/java/com/epmedu/animeal/favourites/presentation/viewmodel/FavouritesViewModel.kt:
```diff
@@ -0,0 +1,77 @@
+            snapshot += if (event.id == feedingPoint.id) {
+                feedingPoint.copy(isFavourite = event.isFavorite)
+            } else {
+                feedingPoint.copy()
```
Redundant `copy()`

ArtemKleinschmidt committed:
Code review comments fix

ArtemKleinschmidt committed:
Merge remote-tracking branch 'origin/develop' into feature/EPMEDU-968_open_the_feeding_point_details_from_the_favorites_page

ArtemKleinschmidt committed:
Code smell fix
